File: Rust/src/donp/component/message.rs
/*
	Package 'component' provides definitions for various objects meeting the definition of a protocol component.
*/

/*
    Objects of type Message represent a specific message within a protocol.

    Required fields: name.
    Other fields are stored in a dictionary corresponding with protocol elements.
*/

use log::{debug};

const INT16: &str = "int16";
const INT32: &str = "int32";
const FLOAT: &str = "float";
const STRING: &str = "string";
const BIT: &str = "bit";

type DataType = String;


pub struct Message{
    name: String,
    message: Vec<u8>,
    msg_dict: std::collections::HashMap<String, serde_json::Value>,    
}

impl Default for Message {
    fn default() -> Self {
        Message {
            name: String::new(),            
            message: Vec::new(),
            msg_dict: std::collections::HashMap::new(),
        }
    }
}

impl Clone for Message {
    fn clone(&self) -> Message {
        Message {
            name: self.name.clone(),
            message: self.message.clone(),
            msg_dict: self.msg_dict.clone(),
        }
    }
}

impl Message {
    pub fn new() -> Message {
        Message::default() 
    }

    pub fn from_map(&mut self, map: &serde_json::Value) {
        // Initialize message fields based on map.
        self.name = map["name"].as_str().unwrap_or("").to_string();
        let map = match map.as_object() {
            Some(m) => m,
            None => return,
        };
        for (key, value) in map {
            if key != "name" {
                self.msg_dict.insert(key.clone(), value.clone());
            }
        }
    }

    pub fn get_name(&self) -> &String {
        &self.name
    }

    pub fn get_message(&self) -> &Vec<u8> {
        &self.message
    }

    pub fn get_msg_dict(&self) -> &std::collections::HashMap<String, serde_json::Value> {
        &self.msg_dict
    }

    pub fn set_message(&mut self, msg: &Vec<u8>) {
        // Set message only if current message is empty and supplied message is not empty.
        if self.message.len() == 0 && msg.len() > 0 {
            self.message = msg.clone();
            return;
        }
    }

    pub fn get_data_byte_count(&self) -> u32 {
        // Determine the byte count based on data type and length.
        let length = self.get_data_length();
        let mut data_type = DataType::from(INT16);
        if self.msg_dict.contains_key("data_type") {
            data_type = self.msg_dict["data_type"].as_str().unwrap_or(INT16).to_string();
        }
        let count = match data_type.to_lowercase().as_str() {
            BIT => if length % 8 == 0 { length / 8 } else { length / 8 + 1 },
            INT16 => length * 2,
            INT32 => length * 4,
            FLOAT => length * 4,
            STRING => length * 1,
            _ => return 0
        };
        count            
    }

    fn get_data_length(&self) -> u32 {
        if self.msg_dict.contains_key("length") {
            return self.msg_dict["length"].as_u64().unwrap_or(0) as u32;
        }
        0
    }

    pub fn log(&self) {
        // Log message information for debugging purposes.
        debug!("Message Name: {}", self.name);
        for (key, value) in &self.msg_dict {
            debug!("Message Dictionary Item: {} => {}", key, value);
        }
    }
}

File: C++/pkg/component/message.cpp
/*
	Package 'component' provides definitions for various objects meeting the definition of a protocol component.
*/

/*
    Objects of type Message represent a specific message within a protocol.

    Required fields: name.
    Other fields are stored in a dictionary corresponding with protocol elements.
*/


#include "pkg/component/message.hpp"

// Constructor
Message::Message() {
    name = "";
    message = VecU8();
    msg_dict = std::map<std::string, std::variant<std::string, uint32_t>>();
    logger = nullptr;
}

Message::Message(std::string json_config, std::shared_ptr<Logger> log_ptr) : logger(log_ptr) {
    // Initialize message from JSON configuration
    auto view = glz::get_view_json<"/name">(json_config);
    if (view) {
        name = std::string(view->begin() + 1, view->end() - 1);		
    }    
    std::map<std::string, std::variant<std::string, uint32_t>> data;
    auto result = glz::read_json(data, json_config);
    if (result.count > 0) {
        for (const auto& [key, value] : data) {
            msg_dict[key] = value;
        }
    } else {
        logger->log(ERROR, "Message: Failed to parse message dictionary from JSON.");
    }
}

// Destructor
Message::~Message() {
    // Cleanup code if needed
}

// Methods
std::string Message::get_name(void) const {
    return name;
}   

VecU8 Message::get_message() const {
    return message;
}

std::map<std::string, std::variant<std::string, uint32_t>> Message::get_msg_dict(void) const{
    return msg_dict;
}

void Message::set_message(const VecU8& msg) {
    // Set mesage only if it is currently empty and supplied msg is not empty.
    if (message.size() == 0 && msg.size() > 0) {
        for (const auto& byte : msg) {
            message.push_back(byte);
        }
        return;
    } else {
        logger->log(ERROR, "Message: Cannot set message data. Message is already set or supplied data is empty.");
    }
}

int32_t Message::get_data_byte_count(void) const {
    // Determine the byte count based on data type and length.
    int32_t length = get_data_length();
    std::string type = DataType(INT16).type; // Default data type
    auto it = msg_dict.find("data_type");
    if (it != msg_dict.end()) {
        type = std::get<std::string>(it->second);
    }
    if (type == INT16) {
        return length * 2;
    } else if (type == INT32 || type == FLOAT) {
        return length * 4;
    } else if (type == STRING) {
        return length;
    } else if (type == BIT) {
        return length % 8 == 0 ? length / 8 : (length / 8) + 1;
    }
    return 0;
}

void Message::log(void) const {
    logger->log(DEBUG, "Message Name: " + name); 
    for (const auto& [key, value] : msg_dict) {
        std::string value_str;
        if (std::holds_alternative<std::string>(value)) {
            value_str = std::get<std::string>(value);
        } else if (std::holds_alternative<uint32_t>(value)) {
            value_str = std::to_string(std::get<uint32_t>(value));
        } 
        logger->log(DEBUG, "  " + key + ": " + value_str);
    }
}

int32_t Message::get_data_length() const {
    // Retrieve the data length from the message dictionary.
    auto it = msg_dict.find("length");
    if (it != msg_dict.end()) {
        return int32_t(std::get<uint32_t>(it->second));
    }
    return int32_t(1);
}

